src/reanimated2/layoutReanimation/sharedTransitions/SharedElementTransition.ts

```diff
@@ -0,0 +1,301 @@
+  unregisterEventHandler,
+} from '../../core';
+import { Platform } from 'react-native';
+import { runOnUIImmediately } from '../../threads';
+import './ProgressTransitionManager';
+
+const supportedProps = [
+  'width',
+  'height',
+  'originX',
+  'originY',
+  'transform',
+  'borderRadius',
+];
+
+type AnimationFactory = (
+  values: SharedTransitionAnimationsValues
+) => StyleProps;
+
+type TransitionProgressEvent = {
+  closing: number;
+  goingForward: number;
+  eventName: string;
+  progress: number;
+  target: number;
+};
+
+const dummyWorklet = () => {
+  'worklet';
+  // noop
+};
+
+export class SharedElementTransition {
+  private _customAnimationFactory: AnimationFactory | null = null;
+  private _animation: SharedTransitionAnimationsFunction | null = null;
+  private _transitionDuration = 500;
+  private _customProgressAnimation?: ProgressAnimation = undefined;
+  private _progressAnimation?: ProgressAnimation = undefined;
+  private _defaultTransitionType?: SharedTransitionType = undefined;
+  private static _sharedElementCount = 0;
+  private static _eventHandler = {
+    isRegistered: false,
+    onTransitionProgress: -1,
+    onAppear: -1,
+    onDisappear: -1,
+    onSwipeDismiss: -1,
+  };
+
+  public animation(
+    customAnimationFactory: AnimationFactory
+  ): SharedElementTransition {
+    this._customAnimationFactory = customAnimationFactory;
+    return this;
+  }
+
+  public progressAnimation(
+    progressAnimationCallback: CustomProgressAnimation
+  ): SharedElementTransition {
+    this._customProgressAnimation = (viewTag, values, progress) => {
+      'worklet';
+      const newStyles = progressAnimationCallback(values, progress);
+      _notifyAboutProgress(viewTag, newStyles, true);
+    };
+    return this;
+  }
+
+  public duration(duration: number): SharedElementTransition {
+    this._transitionDuration = duration;
+    return this;
+  }
+
+  public defaultTransitionType(
+    transitionType: SharedTransitionType
+  ): SharedElementTransition {
+    this._defaultTransitionType = transitionType;
+    return this;
+  }
+
+  public registerTransition(
+    viewTag: number,
+    sharedTransitionTag: string
+  ): void {
+    const transitionAnimation = this.getTransitionAnimation();
+    const progressAnimation = this.getProgressAnimation();
+    if (!this._defaultTransitionType) {
+      if (this._customAnimationFactory && !this._customProgressAnimation) {
+        this._defaultTransitionType = SharedTransitionType.ANIMATION;
+      } else {
+        this._defaultTransitionType = SharedTransitionType.PROGRESS_ANIMATION;
+      }
+    }
+    configureLayoutAnimations(
+      viewTag,
+      LayoutAnimationType.SHARED_ELEMENT_TRANSITION,
+      transitionAnimation,
+      sharedTransitionTag
+    );
+    configureLayoutAnimations(
+      viewTag,
+      LayoutAnimationType.SHARED_ELEMENT_TRANSITION_PROGRESS,
+      dummyWorklet,
+      this._defaultTransitionType
+    );
+    runOnUIImmediately(() => {
```

Comment:
we does it need to be immediate? Do we also use immediate calls for similar purposes elsewhere?